# 라이브러리 임포트 및 데이터 읽어오기

In [19]:
import pandas as pd
import numpy as np
from dataset import Dataset
from dataread import Dataread
import seaborn as sns
import matplotlib.pyplot as plt

sns.set_style("whitegrid")
plt.rcParams['font.family'] = 'NanumGothic'
plt.rcParams['axes.unicode_minus'] = False


ds = Dataset()
dr = Dataread()
JSON_KEY_PATH = "./sprintda05-soomin.json"

In [20]:
master_table = pd.read_csv('./dataset/master_table_v2.csv')
point_table = pd.read_csv('./dataset/use_point_within_signup_N_days.csv')
candidate_table = pd.read_csv('./dataset/candidate_chosen_ratio.csv')

# 테이블 분석하기

In [21]:
def cohort_cut(df):
    over50_cohort = df.loc[df['retention_day8'] >= 0.50]
    under50_cohort = df.loc[df['retention_day8'] < 0.50]
    return over50_cohort, under50_cohort

In [22]:
def result_print(df, func):
    over50_cohort, under50_cohort = cohort_cut(df)

    over50_분석 = func(over50_cohort)
    under50_분석 = func(under50_cohort)
    
    print('Over50')
    display(over50_분석)
    print('='*100)
    print('Under50')
    display(under50_분석)
    return

In [23]:
master_table

user_id  total_votes  unique_days  first_vote   last_vote  active_days  \
0      838023           22            5  2023-04-29  2023-05-04            5   
1      838466          127            6  2023-05-02  2023-05-07            5   
2      838642            1            1  2023-04-29  2023-04-29            0   
3      839357          205            8  2023-04-29  2023-05-07            8   
4      840293          170            7  2023-05-03  2023-05-15           12   
...       ...          ...          ...         ...         ...          ...   
4698  1510561            9            1  2023-05-25  2023-05-25            0   
4699  1511957            8            1  2023-05-25  2023-05-25            0   
4700  1512422           22            1  2023-05-25  2023-05-25            0   
4701  1512542            7            1  2023-05-25  2023-05-25            0   
4702  1513236            7            1  2023-05-25  2023-05-25            0   

      cohort_day  retention_day8 ret_group ret_group_over ret_group_under  \
0     2023-04-29        0.515152     50 이상          50 이상             NaN   
1     2023-05-02        0.569444     50 이상          50 이상             NaN   
2     2023-04-29        0.515152     50 이상          50 이상             NaN   
3     2023-04-29        0.515152     50 이상          50 이상             NaN   
4     2023-05-03        0.422977     50 미만            NaN           50 미만   
...          ...             ...       ...            ...             ...   
4698  2023-05-25        0.451613     50 미만            NaN           50 미만   
4699  2023-05-25        0.451613     50 미만            NaN           50 미만   
4700  2023-05-25        0.451613     50 미만            NaN           50 미만   
4701  2023-05-25        0.451613     50 미만            NaN           50 미만   
4702  2023-05-25        0.451613     50 미만            NaN           50 미만   

     gender  group_id  school_id  grade  
0         F     233.0      271.0    1.0  
1         F     321.0      369.0    2.0  
2         F     356.0      271.0    2.0  
3         M     519.0      271.0    1.0  
4         F     838.0      369.0    1.0  
...     ...       ...        ...    ...  
4698      M   73885.0     4426.0    2.0  
4699      F   29558.0     5520.0    2.0  
4700      F   72313.0     4426.0    2.0  
4701      F   72313.0     4426.0    2.0  
4702      F   70559.0     4426.0    3.0  

[4703 rows x 15 columns]

## candidate, point 분석

In [24]:
user_table = master_table[['user_id','total_votes','unique_days','retention_day8','gender','school_id','grade']]
user_table = user_table.merge(point_table, on='user_id',how='left')
user_table = user_table.merge(candidate_table, on='user_id', how='left')
user_table

user_id  total_votes  unique_days  retention_day8 gender  school_id  \
0      838023           22            5        0.515152      F      271.0   
1      838466          127            6        0.569444      F      369.0   
2      838642            1            1        0.515152      F      271.0   
3      839357          205            8        0.515152      M      271.0   
4      840293          170            7        0.422977      F      369.0   
...       ...          ...          ...             ...    ...        ...   
4698  1510561            9            1        0.451613      M     4426.0   
4699  1511957            8            1        0.451613      F     5520.0   
4700  1512422           22            1        0.451613      F     4426.0   
4701  1512542            7            1        0.451613      F     4426.0   
4702  1513236            7            1        0.451613      F     4426.0   

      grade                sign_up_date      first_vote_date  \
0       1.0  2023-04-19 09:06:00.719792  2023-04-29 16:22:56   
1       2.0  2023-04-19 15:10:56.560337  2023-05-02 22:42:05   
2       2.0  2023-04-20 00:11:00.962160  2023-04-28 14:18:07   
3       1.0  2023-04-20 14:26:20.479856  2023-04-29 16:51:01   
4       1.0  2023-04-21 11:45:56.750174  2023-05-03 02:24:05   
...     ...                         ...                  ...   
4698    2.0  2023-05-25 14:27:02.976395  2023-05-25 14:28:20   
4699    2.0  2023-05-25 15:45:50.814710  2023-05-25 15:46:25   
4700    2.0  2023-05-25 16:52:43.081341  2023-05-25 16:57:53   
4701    2.0  2023-05-25 20:03:38.172395  2023-05-25 22:59:47   
4702    3.0  2023-05-25 23:21:21.182921  2023-05-25 23:22:21   

     first_point_use_date  first_point_use_amount  vote_sameday_signup  \
0     2023-04-30 14:27:35                  -300.0                    0   
1     2023-05-05 07:32:34                  -300.0                    0   
2     2023-04-30 06:49:01                  -300.0                    0   
3     2023-04-30 02:19:41                  -300.0                    0   
4     2023-05-05 14:49:11                  -300.0                    0   
...                   ...                     ...                  ...   
4698  2023-05-26 15:24:26                  -200.0                    1   
4699                  NaN                     NaN                    1   
4700  2023-05-25 17:09:17                   -10.0                    1   
4701  2023-05-26 11:28:52                  -200.0                    1   
4702  2023-05-26 11:50:06                   -10.0                    1   

      point_used_within_signup_3days  candidate_count  chosen_user_id  \
0                                  0             27.0        838023.0   
1                                  0             82.0        838466.0   
2                                  0             44.0        838642.0   
3                                  0             42.0        839357.0   
4                                  0             74.0        840293.0   
...                              ...              ...             ...   
4698                               1              NaN             NaN   
4699                               0              NaN             NaN   
4700                               1              NaN             NaN   
4701                               1              NaN             NaN   
4702                               1              NaN             NaN   

      chosen_count     ratio  
0             10.0  0.370370  
1             37.0  0.451220  
2              5.0  0.113636  
3             10.0  0.238095  
4             31.0  0.418919  
...            ...       ...  
4698           NaN       NaN  
4699           NaN       NaN  
4700           NaN       NaN  
4701           NaN       NaN  
4702           NaN       NaN  

[4703 rows x 17 columns]

In [25]:
def candidate_analysis(df):
    result = df.agg({
    'first_point_use_amount': 'mean',
    'vote_sameday_signup': 'mean',
    'candidate_count': 'sum',
    'chosen_count': 'sum',
    'total_votes':'sum',
    'unique_days':'sum',
    'grade':'mean'
})
    result['chosen_ratio'] = result['chosen_count'] / result['candidate_count']
    result['vote_per_day'] = result['total_votes'] / result['unique_days']

    return result.reset_index()

In [26]:
result_print(user_table, candidate_analysis)

Over50


index              0
0  first_point_use_amount    -272.830084
1     vote_sameday_signup       0.778684
2         candidate_count   99609.000000
3            chosen_count   33720.000000
4             total_votes  457737.000000
5             unique_days   15878.000000
6                   grade       1.824905
7            chosen_ratio       0.338524
8            vote_per_day      28.828379

Under50


index              0
0  first_point_use_amount    -273.762963
1     vote_sameday_signup       0.822626
2         candidate_count  121998.000000
3            chosen_count   41175.000000
4             total_votes  554561.000000
5             unique_days   20280.000000
6                   grade       1.890014
7            chosen_ratio       0.337506
8            vote_per_day      27.345217

#### 해석
- point_use_amount -300 << 첫 포인트가 투표 선택된 것에 사용한 서비스
- point_use_amount -200 << 사용처를 알 수 없음. 선택받은 적 없는 사람도 사용 가능한 포인트 서비스
- 따라서 -300에 가까울수록 투표 관련하여 포인트를 사용했다고 볼 수 있다. 그러나 유의미한 차이가 없음.
    

- vote_sameday_signup << 가입 첫 날 바로 활성화된 비율. 77% vs 82% 인 상황. 약간의 차이가 있기는 하지만 유의미한 차이라고 보이진 않음.


- ratio << 후보 등장 대비 선택받은 비율. 두 집단 모두 33% 로 사실상 무차별한 것으로 보임.

## VOTE POINT within after N days

In [27]:
import warnings
# SettingWithCopyWarning 무시
warnings.simplefilter(action='ignore', category=pd.errors.SettingWithCopyWarning)

In [28]:
user_table = master_table[['user_id','total_votes','unique_days','retention_day8','gender','school_id','grade']]
user_table = user_table.merge(point_table, on='user_id',how='left')
user_table = user_table.merge(candidate_table, on='user_id', how='left')
user_table

user_id  total_votes  unique_days  retention_day8 gender  school_id  \
0      838023           22            5        0.515152      F      271.0   
1      838466          127            6        0.569444      F      369.0   
2      838642            1            1        0.515152      F      271.0   
3      839357          205            8        0.515152      M      271.0   
4      840293          170            7        0.422977      F      369.0   
...       ...          ...          ...             ...    ...        ...   
4698  1510561            9            1        0.451613      M     4426.0   
4699  1511957            8            1        0.451613      F     5520.0   
4700  1512422           22            1        0.451613      F     4426.0   
4701  1512542            7            1        0.451613      F     4426.0   
4702  1513236            7            1        0.451613      F     4426.0   

      grade                sign_up_date      first_vote_date  \
0       1.0  2023-04-19 09:06:00.719792  2023-04-29 16:22:56   
1       2.0  2023-04-19 15:10:56.560337  2023-05-02 22:42:05   
2       2.0  2023-04-20 00:11:00.962160  2023-04-28 14:18:07   
3       1.0  2023-04-20 14:26:20.479856  2023-04-29 16:51:01   
4       1.0  2023-04-21 11:45:56.750174  2023-05-03 02:24:05   
...     ...                         ...                  ...   
4698    2.0  2023-05-25 14:27:02.976395  2023-05-25 14:28:20   
4699    2.0  2023-05-25 15:45:50.814710  2023-05-25 15:46:25   
4700    2.0  2023-05-25 16:52:43.081341  2023-05-25 16:57:53   
4701    2.0  2023-05-25 20:03:38.172395  2023-05-25 22:59:47   
4702    3.0  2023-05-25 23:21:21.182921  2023-05-25 23:22:21   

     first_point_use_date  first_point_use_amount  vote_sameday_signup  \
0     2023-04-30 14:27:35                  -300.0                    0   
1     2023-05-05 07:32:34                  -300.0                    0   
2     2023-04-30 06:49:01                  -300.0                    0   
3     2023-04-30 02:19:41                  -300.0                    0   
4     2023-05-05 14:49:11                  -300.0                    0   
...                   ...                     ...                  ...   
4698  2023-05-26 15:24:26                  -200.0                    1   
4699                  NaN                     NaN                    1   
4700  2023-05-25 17:09:17                   -10.0                    1   
4701  2023-05-26 11:28:52                  -200.0                    1   
4702  2023-05-26 11:50:06                   -10.0                    1   

      point_used_within_signup_3days  candidate_count  chosen_user_id  \
0                                  0             27.0        838023.0   
1                                  0             82.0        838466.0   
2                                  0             44.0        838642.0   
3                                  0             42.0        839357.0   
4                                  0             74.0        840293.0   
...                              ...              ...             ...   
4698                               1              NaN             NaN   
4699                               0              NaN             NaN   
4700                               1              NaN             NaN   
4701                               1              NaN             NaN   
4702                               1              NaN             NaN   

      chosen_count     ratio  
0             10.0  0.370370  
1             37.0  0.451220  
2              5.0  0.113636  
3             10.0  0.238095  
4             31.0  0.418919  
...            ...       ...  
4698           NaN       NaN  
4699           NaN       NaN  
4700           NaN       NaN  
4701           NaN       NaN  
4702           NaN       NaN  

[4703 rows x 17 columns]

### VOTE

In [29]:
date_user_table = user_table[['user_id','retention_day8','sign_up_date','first_vote_date','first_point_use_date']]
date_user_table['sign_up_date'] = pd.to_datetime(date_user_table['sign_up_date'])
date_user_table['first_vote_date'] = pd.to_datetime(date_user_table['first_vote_date'])
date_user_table['first_point_use_date'] = pd.to_datetime(date_user_table['first_point_use_date'])

In [30]:
# 조건 컬럼
date_user_table['days_to_vote'] = (date_user_table['first_vote_date'] - date_user_table['sign_up_date']).dt.days

# 조건별로 각 컬럼 생성
for i in range(1, 7):
    date_user_table[f'vote_within_{i}days'] = (date_user_table['days_to_vote'] <= i).astype(int)

# 7일 초과 투표
date_user_table['vote_over_7days'] = (date_user_table['days_to_vote'] > 7).astype(int)

In [31]:
def within_days_analysis(df):
    df = df.drop(columns=['user_id','retention_day8','sign_up_date','first_vote_date','first_point_use_date'])
    result = df.mean()
    return result.reset_index()

In [32]:
result_print(date_user_table, within_days_analysis)

Over50


index         0
0       days_to_vote  0.542142
1  vote_within_1days  0.899946
2  vote_within_2days  0.929853
3  vote_within_3days  0.942904
4  vote_within_4days  0.960848
5  vote_within_5days  0.964111
6  vote_within_6days  0.977162
7    vote_over_7days  0.019032

Under50


index         0
0       days_to_vote  0.272696
1  vote_within_1days  0.952514
2  vote_within_2days  0.969623
3  vote_within_3days  0.979050
4  vote_within_4days  0.984986
5  vote_within_5days  0.988128
6  vote_within_6days  0.990223
7    vote_over_7days  0.008380

#### 해석
- 전반적으로 under50 이 가입대비 더 이른시점에 투표를 진행하는 것으로 보인다.
- 이렇게 된 이후는, signup day 가 4.19일 등의 시점이지만, 실제 투표가 기록되는 날짜는 4.28일 이후임에 따라서    
4.28~5.1에 해당하는 초반부 코호트가 over50 코호트로 잡히면서 발생한 문제가 아닐까 싶다.

- 가장 대표적으로 확인할 수 있는게, 첫 투표까지 걸린 기간이 0.54일인데 비해, under50 코호트에서 0.27일로 4명중 3명은 회원 가입 당일에 바로 투표 서비스를 이용했다 처럼 보이는 것.

### POINT

In [33]:
## 포인트 기록

date_user_table = user_table[['user_id','retention_day8','sign_up_date','first_vote_date','first_point_use_date']]
date_user_table['sign_up_date'] = pd.to_datetime(date_user_table['sign_up_date'])
date_user_table['first_point_use_date'] = pd.to_datetime(date_user_table['first_point_use_date'])

# 조건 컬럼
date_user_table['days_to_point'] = (date_user_table['first_point_use_date'] - date_user_table['sign_up_date']).dt.days

# 조건별로 각 컬럼 생성
for i in range(1, 7):
    date_user_table[f'point_within_{i}days'] = (date_user_table['days_to_point'] <= i).astype(int)

# 7일 초과 투표
date_user_table['point_over_7days'] = (date_user_table['days_to_point'] > 7).astype(int)

result_print(date_user_table, within_days_analysis)

Over50


index         0
0       days_to_point  1.335933
1  point_within_1days  0.793910
2  point_within_2days  0.846112
3  point_within_3days  0.874932
4  point_within_4days  0.895595
5  point_within_5days  0.909190
6  point_within_6days  0.928222
7    point_over_7days  0.037520

Under50


index         0
0       days_to_point  1.064444
1  point_within_1days  0.800279
2  point_within_2days  0.854749
3  point_within_3days  0.879190
4  point_within_4days  0.895251
5  point_within_5days  0.905377
6  point_within_6days  0.913408
7    point_over_7days  0.027584

#### 해석
- SIGNUP DATE 가 기준이 되었기 때문에, 역시나 POINT 사용 역시 OVER50 코호트에서 평균적으로 좀 더 지연된 경향이 있다.

### POINT USE AFTER VOTE

In [34]:
## VOTE 대비 POINT TIME
date_user_table = user_table[['user_id','retention_day8','sign_up_date','first_vote_date','first_point_use_date']]
date_user_table['first_vote_date'] = pd.to_datetime(date_user_table['first_vote_date'])
date_user_table['first_point_use_date'] = pd.to_datetime(date_user_table['first_point_use_date'])

# 조건 컬럼
date_user_table['days_to_point_after_vote'] = (date_user_table['first_point_use_date'] - date_user_table['first_vote_date']).dt.days

# 조건별로 각 컬럼 생성
for i in range(1, 7):
    date_user_table[f'point_within_{i}vote_days'] = (date_user_table['days_to_point_after_vote'] <= i).astype(int)

# 7일 초과 투표
date_user_table['point_over_7vote_days'] = (date_user_table['days_to_point_after_vote'] > 7).astype(int)

result_print(date_user_table, within_days_analysis)

Over50


index         0
0  days_to_point_after_vote  0.657939
1   point_within_1vote_days  0.878738
2   point_within_2vote_days  0.912996
3   point_within_3vote_days  0.934747
4   point_within_4vote_days  0.944535
5   point_within_5vote_days  0.952692
6   point_within_6vote_days  0.955954
7     point_over_7vote_days  0.016313

Under50


index         0
0  days_to_point_after_vote  0.674444
1   point_within_1vote_days  0.849860
2   point_within_2vote_days  0.888268
3   point_within_3vote_days  0.900838
4   point_within_4vote_days  0.913059
5   point_within_5vote_days  0.918296
6   point_within_6vote_days  0.921439
7     point_over_7vote_days  0.020251

#### 해석
- SIGNUP이 문제라면, VOTE 대비로 봐보면 다를까?
    - 이전 결과와 달리, 평균 0.66일로 두 코호트간 차이가 거의 없는것으로 나타난다.
    - 다만 세부적으로, Over50 그룹이 조금 더 투표날 대비 포인트 사용날이 더 빠른것으로 추정된다.

### 추가 방향성
    잘 모르겠다~
